stepflow-rs/crates/stepflow-execution/src/debug_executor.rs

```diff
@@ -0,0 +1,1177 @@
+                let status = if ready.contains(idx) {
+                    StepStatus::Runnable
+                } else {
+                    StepStatus::Blocked
+                };
+                queued.push(StepExecution::new(
+                    idx,
+                    step.id.clone(),
+                    step.component.to_string(),
+                    status,
+                ));
+            }
+        }
+
+        queued
+    }
+
+    /// Get the cached result of a step, or None if not yet completed.
+    pub fn get_step_result(&self, step_id: &str) -> Option<FlowResult> {
+        let step_index = self.state.step_index_map().step_index(step_id)?;
+        self.state.get_step_result(step_index).cloned()
+    }
+
+    // ========================================================================
+    // Debug Queue Persistence
+    // ========================================================================
+
+    /// Add steps to the persistent debug queue.
+    ///
+    /// This should be called after `queue_step()` returns newly queued steps.
+    /// The queue persists across HTTP requests for step-through debugging.
+    pub async fn add_steps_to_debug_queue(&self, step_ids: &[String]) -> Result<()> {
+        if step_ids.is_empty() {
+            return Ok(());
+        }
+
+        self.state_store
+            .add_to_debug_queue(self.context.run_id(), step_ids)
+            .await
+            .change_context(ExecutionError::StateError)?;
+
+        log::debug!(
+            "Added to debug queue for run {}: [{}]",
+            self.context.run_id(),
+            step_ids.join(", ")
+        );
+
+        Ok(())
+    }
+
+    /// Remove steps from the persistent debug queue.
+    ///
+    /// This should be called after executing steps via `run_next_step()` or `run_queue()`.
+    pub async fn remove_steps_from_debug_queue(&self, step_ids: &[String]) -> Result<()> {
+        if step_ids.is_empty() {
+            return Ok(());
+        }
+
+        self.state_store
+            .remove_from_debug_queue(self.context.run_id(), step_ids)
+            .await
+            .change_context(ExecutionError::StateError)?;
+
+        log::debug!(
+            "Removed from debug queue for run {}: [{}]",
+            self.context.run_id(),
+            step_ids.join(", ")
+        );
+
+        Ok(())
+    }
+
+    // ========================================================================
+    // Recovery (internal)
+    // ========================================================================
+
+    /// Recover execution state from the state store and fix any missed status transitions.
+    ///
+    /// This is called internally by [`resume()`](Self::resume) to restore completed steps
+    /// and fix any status mismatches.
+    ///
+    /// # Returns
+    /// The number of step status corrections made during recovery
+    async fn recover_from_state_store(&mut self) -> Result<usize> {
```

Comment:
This should be two parts -- recreating the `DebugState` from the state store, and then continuing the execution as needed.